Add commits dropdown in PR files view and allow commit by commit review

sebastian-sauer opened the pull request:
This PR adds a new dropdown to select a commit or a commit range (shift-click like github) of a Pull Request. 
After selection of a commit only the changes of this commit will be shown. When selecting a range of commits the diff of this range is shown.

This allows to review a PR commit by commit or by viewing only commit ranges.

The "Show changes since your last review" mechanism github uses is implemented, too.

When reviewing a single commit or a commit range the "Viewed" functionality is disabled.

## Screenshots

### The commit dropdown
![image](https://github.com/go-gitea/gitea/assets/51889757/0db3ae62-1272-436c-be64-4730c5d611e3)

### Selecting a commit range
![image](https://github.com/go-gitea/gitea/assets/51889757/ad81eedb-8437-42b0-8073-2d940c25fe8f)

### Show changes of a single commit only
![image](https://github.com/go-gitea/gitea/assets/51889757/6b1a113b-73ef-4ecc-adf6-bc2340bb8f97)

### Show changes of a commit range
![image](https://github.com/go-gitea/gitea/assets/51889757/6401b358-cd66-4c09-8baa-6cf6177f23a7)


Fixes https://github.com/go-gitea/gitea/issues/20989
Fixes https://github.com/go-gitea/gitea/issues/19263

delvh commented:
Hmm…
I wonder how that will play with #19007.
I can foresee at least three bugs that will become more present compared to now:
- #24813
- The architecture is built at the moment to use the **newest** review as the latest review to compare against.
As long as you only move forward in the time, this is not a problem.
There's no performant way around this issue, as it would otherwise mean a **very** expensive comparison what the latest commit a review happened on is.
With your proposed approach, we can also backtrack in time, potentially leading to comparing the state against an older one than necessary.
- The other way around is most likely true as well (didn't test it):
If you review a PR on the newest file, and then switch to an older commit, files you haven't viewed yet will probably be marked as viewed.

I think for now, the easiest thing to combat all these problems is to **disable** the `viewed` checkboxes (at best with a tooltip explaining why) (or alternatively hiding it completely) when you are only reviewing a specific commit of the PR.

sebastian-sauer commented:
> Hmm… I wonder how that will play with #19007. I can foresee at least three bugs that will become more present compared to now:
> 
> * [Files are viewed on too new commits #24813](https://github.com/go-gitea/gitea/issues/24813)
> * The architecture is built at the moment to use the **newest** review as the latest review to compare against.
>   As long as you only move forward in the time, this is not a problem.
>   There's no performant way around this issue, as it would otherwise mean a **very** expensive comparison what the latest commit a review happened on is.
>   With your proposed approach, we can also backtrack in time, potentially leading to comparing the state against an older one than necessary.
> * The other way around is most likely true as well (didn't test it):
>   If you review a PR on the newest file, and then switch to an older commit, files you haven't viewed yet will probably be marked as viewed.
> 
> I think for now, the easiest thing to combat all these problems is to **disable** the `viewed` checkboxes (at best with a tooltip explaining why) (or alternatively hiding it completely) when you are only reviewing a specific commit of the PR.

Thanks for your feedback - disabling the Viewed Checkboxes should be an easy thing. I'll add this.
... by the way this reminds me to analyse the missing headcommit infomation ;-)

silverwind commented:
As for UI:

- I don't think radio button is the right. If at all, it should have a checkbox. 
- It would be nice to aallow selection of unbroken commit ranges, which will require a bit of JS to get working.
- There needs to be a border between dropdown elements
- All text-holding elements need to ellipsis-wrap, there is a helper class for it.

sebastian-sauer commented:
thanks @silverwind - do you mean all span (class description and text) should have the class gt-ellipsis?

silverwind commented:
At least the commit message and author line will needed it. Commit hash is fixed width, so doesn't.

sebastian-sauer commented:
> At least the commit message and author line will needed it. Commit hash is fixed width, so doesn't.

Thanks - i've pushed an updated version:

![image](https://github.com/go-gitea/gitea/assets/1135157/2313f168-b322-4a82-9fc5-5538a1f4f1a4)

sebastian-sauer commented:
@silverwind are there any more ui / review comments that I'll need to address? 

@delvh is the current solution (disabling file viewed when loading specific commit) ok for you?

silverwind commented:
Looks good. Something to test:

What happens when number of commits is large. Will the dropdown become scrollable?

delvh commented:
> @delvh is the current solution (disabling file viewed when loading specific commit) ok for you?

Yep, of course. Otherwise I wouldn't have suggested it.

silverwind commented on web_src/css/repo.css:
```diff
@@ -3344,4 +3344,9 @@ tbody.commit-list {
+.commit-selector-menu {
+  max-height: 60vh;
+  overflow-y: auto;
 }
```
```suggestion
}

```

silverwind commented:
@sebastian-sauer install a https://editorconfig.org/ plugin in your editor.

sebastian-sauer commented:
Thanks @silverwind - i've now added a max height and overflow-y auto to add scrollbar for long lists.

The following screenshot is with 500 commits in one PR:

![image](https://github.com/go-gitea/gitea/assets/1135157/17f9c224-1e84-4d22-9b87-045bdbf313f3)

silverwind commented:
I guess we can reduce item padding a bit in that dropdown specifically, maybe `padding: 6px 3px`.

puni9869 commented:
> padding: 6px 3p

I was thinking the same working on it

puni9869 commented:
Apologies,  
I wrongly put the `#`. Github is trying to be oversmart some time. 😅 
![image](https://github.com/go-gitea/gitea/assets/80308335/b7380689-3f1b-4bc3-b62c-1bcd7ceedbcf)

sebastian-sauer commented:
I've reduced padding and the margin of the divider:

![image](https://github.com/go-gitea/gitea/assets/1135157/2a034675-4a11-43de-9d13-53297abf5706)

silverwind commented:
Bit too smal actually, I will check out the branch and correct it, before making any more misguided requests :D

silverwind commented:
Did a number of UI enhancements. Looks much better to me now:

<img width="715" alt="Screenshot 2023-06-28 at 22 41 51" src="https://github.com/go-gitea/gitea/assets/115237/31e13413-e41f-4ff3-a3ed-cae32309ef05">

silverwind reviewed (approved):
UI looking good. A backend integration test might be nice to have.

silverwind commented on web_src/css/base.css:
```diff
@@ -630,6 +630,11 @@ a.label,
+.ui.dropdown .menu > .item > .description {
+  color: var(--color-text-light-2);
+}
+
```
This is not strictly for this PR anymore, but I saw we were missing this color for dark theme, so it's good to have it in any case.

sebastian-sauer commented:
Your support is much appreciated @silverwind - thank you!

silverwind commented:
I wonder if there is any performance import to this on PRs with many commits. I guess not as the commits are already retrieved currently, right?

sebastian-sauer commented:
> I wonder if there is any performance import to this on PRs with many commits. I guess not as the commits are already retrieved currently, right?

Yes - the commits are already loaded in prInfo *git.CompareInfo. So there is no additional loading of data.

sebastian-sauer commented:
Is there anything blocking a second approval / merge?

lunny commented on routers/web/repo/pull.go:
```diff
@@ -705,6 +707,20 @@ func ViewPullFiles(ctx *context.Context) {
+		foundCommit := false
+		for i := range prInfo.Commits {
+			if prInfo.Commits[i].ID.String() == commitToShow {
+				foundCommit = true
```
break?

lunny commented on routers/web/repo/pull.go:
```diff
@@ -732,19 +756,35 @@ func ViewPullFiles(ctx *context.Context) {
 	var methodWithError string
 	var diff *gitdiff.Diff
-	if !ctx.IsSigned {
+	if !ctx.IsSigned || willShowSingleCommit {
```
I cannot understand the logic.

sebastian-sauer replied:
i've added a comment to the code explaining the or

lunny commented on routers/web/web.go:
```diff
@@ -1286,6 +1286,7 @@ func registerRoutes(m *web.Route) {
 			m.Post("/cleanup", context.RepoMustNotBeArchived(), context.RepoRef(), repo.CleanUpPullRequest)
 			m.Group("/files", func() {
 				m.Get("", context.RepoRef(), repo.SetEditorconfigIfExists, repo.SetDiffViewStyle, repo.SetWhitespaceBehavior, repo.SetShowOutdatedComments, repo.ViewPullFiles)
+				m.Get("/{sha}", context.RepoRef(), repo.SetEditorconfigIfExists, repo.SetDiffViewStyle, repo.SetWhitespaceBehavior, repo.SetShowOutdatedComments, repo.ViewPullFiles)
```
How about to use a query parameter, then we don't need to copy the line.

sebastian-sauer replied:
I tried to use the same Uri style as GitHub (so in future also a range of commits could be added)

sebastian-sauer replied:
i just did some more testing with the github urls and will try to align the urls as much as possible

github uses 3 different ways:
- show single commits diff in PR context: /pull/{pull-id}/commits/{sha1}
- show diff of all commits since (of PR) :  /pull/{pull-id}/files/{sha1}
- show diff of a commit range (of PR): /pull/{pull-id}/files/{sha1}..{sha1}

silverwind replied:
Definitely prefer GH-style URLs.

sebastian-sauer replied:
Yes - i just added the routes.

The commit selection dropdown gets more complex now, so i think i'll transfer it to a vue component.

silverwind replied:
I guess that's fine. Generally I'm wary of new Vue stuff because they cause loading artifacts, but given that the dropdown is not rendered on page load, this one is safe to convert.

sebastian-sauer replied:
I'd prefer a reverse logic of the second route (show all changes in PR from start till specified commit id). 
I think the use case is more often "show all changes since this commit" - for example after you did a review and only want to see new files / changes.

sebastian-sauer replied:
But it's ok for me - i'll change the logic for this case.

sebastian-sauer commented on web_src/js/components/DiffCommitSelector.vue:
```diff
@@ -0,0 +1,77 @@
+
+<style scoped>
+  .hovered {
+    background-color: purple !important;
```
@silverwind  which hover / select color would you propose to use in the commit select dropdown?

silverwind replied:
Background hover? `var(--color-hover)` usually.

sebastian-sauer commented:
My tests look good - i've updated PR description and screenshots.

Any feedback / approve / comment is much appreciated!

sebastian-sauer commented:
> UI looking good. A backend integration test might be nice to have.

I've added integration tests for the new urls.

lunny commented:
CI failure is related.

silverwind commented:
Will review again after latest Vue changes.

sebastian-sauer commented:
I've improved the integration tests with a pr that has more files to really ensure the correct files are shown.

Happy to get any feedback.

vgerasimov11 commented:
Is there any chance to have this functionality in 1.20 release?

6543 commented:
no as v1.20 is already in feature freeze and branched out ... so it would take a backport.

backport's are only done fore bug-fixes and trivial enhancements

silverwind commented:
When I try this branch I get a 500 on PR files tab, it seems to be caused by this PR:

```
GetReviewByIssueIDAndUserID, Error 1728 (HY000): Cannot load from mysql.proc. The table is probably corrupted
```

<img width="997" alt="Screenshot 2023-07-07 at 16 55 39" src="https://github.com/go-gitea/gitea/assets/115237/eb1ac9c4-7a49-4ee3-a259-6309a9df8ceb">

silverwind commented:
CI also shows a few 500s, I think they might be the same cause.

silverwind commented on templates/repo/diff/box.tmpl:
```diff
@@ -31,12 +31,57 @@
+									CommitterOrAuthorName: {{if .Committer}}{{.Committer.Name}}{{else}}{{.Author.Name}}{{end}},
+									ID: {{.ID.String}},
+									Time: {{if .Committer}}{{.Committer.When}}{{else}}{{.Author.When}}{{end}},
+									Selected: false
```
Ideally lowercase the JSON keys.

silverwind commented on templates/repo/diff/box.tmpl:
```diff
@@ -31,12 +31,57 @@
+							show_changes_since_your_last_review: {{.locale.Tr "repo.pulls.show_changes_since_your_last_review"}},
+						},
+						queryParams: '?style={{if $.IsSplitStyle}}split{{else}}unified{{end}}&whitespace={{$.WhitespaceBehavior}}&show-outdated={{$.ShowOutdatedComments}}'
+					};
```
@wxiaoguang is this correct usage of `window.config.pageData`? I'm not sure myself.

wxiaoguang replied:
I do not think it's good to put everything in one page request.

Ideally, the "popup" element can do lazy-loading. Use a separate backend handler to provide related data, the frontend code just does "fetch"

silverwind replied:
IIRC, fomantic popup has some way of lazy-loading, but I guess we should not use it because we eventually want to migrate away from fomantic, and this would just problematically deepen our integration against it.

wxiaoguang replied:
This PR's code is not related to Fomantic's JS.

silverwind replied:
No, but how would you implement "popup" element can do lazy-loading"?

wxiaoguang replied:
> Just `fetch` before opening it?

Yes, that's a general approach, simple and clear enough.

sebastian-sauer replied:
yes i understand that only loading the data when needed is a good approach. But in this case all of the information where already available on the page, so i was just using the information in viewPullfiles.

silverwind replied:
If the template block is only loaded where needed and the data is already computed, I see no issue doing it like this because it will be better UX if they don't have to wait for the `fetch`.

sebastian-sauer replied:
Not sure - if you want i can create a new endpoint and fetch the information when opening the component.

What do you think @wxiaoguang ?

wxiaoguang replied:
If all related data is already prepared by the existing backend code, then I am neutral for it (current approach or fetch/lazy-loading)

A pure fetch solution could avoid the fragile `pageData`, but it's not a must at the moment IMO.

sebastian-sauer replied:
i just added a new route and the component now fetches the data.

silverwind commented on web_src/js/components/DiffCommitSelector.vue:
```diff
@@ -0,0 +1,110 @@
+  }
+};
+</script>
+
```
```suggestion
```

silverwind commented:
Overall it looks pretty good to me. Need to sort out the 500, then I will test again.

wxiaoguang commented:
Instead of adding more and more testing fixtures/repositories, could the existing data be reused for testing purpose?

sebastian-sauer commented:
> 

Yep - just debugged this problem... Looks like there is a bug in FindLatestReviews

the max ( ) should be max() for mysql  (without the space)

sebastian-sauer commented:
> Instead of adding more and more testing fixtures/repositories, could the existing data be reused for testing purpose?

I didn't find a good one for these tests with multiple commits in a PR.

sebastian-sauer commented:
@silverwind i'd really like to add a divider / header line in the dropdown like github does:
![image](https://github.com/go-gitea/gitea/assets/1135157/810edcfd-8440-4bd1-9f14-6a0f877993f4)

(but maybe in a second PR) 

Are you aware of any good template / place to "copy" from?

silverwind commented:
Issue label filter has a fake "header" in the dropdown, but I guess it's not the cleanest implementation IIRC.

wxiaoguang commented on templates/repo/diff/box.tmpl:
```diff
@@ -31,12 +31,57 @@
 			{{template "repo/diff/whitespace_dropdown" .}}
 			{{template "repo/diff/options_dropdown" .}}
+			{{if .PageIsPullFiles}}
+				<script id="commit-data-script">
```
What's the `id` used for?

sebastian-sauer replied:
removed it.

wxiaoguang commented on templates/repo/diff/box.tmpl:
```diff
@@ -31,12 +31,57 @@
+								},
+						{{end}}],
+						locale: {
+							lang: {{$.locale}},
```
`{{$.locale}}` doesn't seem right

sebastian-sauer replied:
fixed now - needed info is loaded via fetch

wxiaoguang commented on templates/repo/diff/box.tmpl:
```diff
@@ -31,12 +31,57 @@
+						the following will be replaced by vue component
+						but this avoids any loading artifacts till the vue component is initialized
+					*/}}
+					<div class="ui jump dropdown basic button custom">
```
```suggestion
					<div class="ui dropdown basic button custom">
```

The `jump` has no effect IMO

wxiaoguang commented on templates/repo/diff/box.tmpl:
```diff
@@ -31,12 +31,57 @@
 	{{if not .DiffNotAvailable}}
+		{{if and .IsShowingOnlySingleCommit .PageIsPullFiles}}
+			<div class="ui info message">
+				<div class="text gt-whitespace-pre">{{.locale.Tr "repo.pulls.showing_only_single_commit" (ShortSha .BeforeCommitID)}}</div>
```
Why `<div class="text gt-whitespace-pre">`? The message doesn't seem special

sebastian-sauer replied:
this is / was just a copy paste - i've removed it

wxiaoguang commented on web_src/js/components/DiffCommitSelector.vue:
```diff
@@ -0,0 +1,109 @@
+
+      <template v-for="commit in commits" :key="commit.id">
+        <div class="divider"/>
+        <div class="vertical item gt-df gt-fr gt-gap-2" :class="{selected: commit.selected}" @click.exact="commitClicked(commit.id)" @click.shift.exact.stop.prevent="commitClickedShift(commit)">
```
I really have difficulty to read code with a lot of `gt` patches ....  I can't tell what is `gt-fc` or `gt-fr` every time .....

silverwind replied:
Yes we should refactor to tailwind-compatible classnames eventually, but I don't see a big issue here.

silverwind commented:
> regarding the 500 on mysql:
> 
> Looks like there is a bug in FindLatestReviews (already existing function - introduced in #24796)
> 
> the max ( ) should be max() for mysql (without the space)

would it be worth to backport a fix for this only to 1.20/1.19?

wxiaoguang commented on templates/repo/diff/box.tmpl:
```diff
@@ -31,12 +31,38 @@
 			{{template "repo/diff/options_dropdown" .}}
+			{{if .PageIsPullFiles}}
+				<script>
+					window.config.pageData.commitInfo = {
+						issueLink: {{$.Issue.Link}},
+						queryParams: '?style={{if $.IsSplitStyle}}split{{else}}unified{{end}}&whitespace={{$.WhitespaceBehavior}}&show-outdated={{$.ShowOutdatedComments}}'
+					};
```
I think there is no need to write the `<script>` now. The `issueLink` and `queryParams` can be put on the `<div>` by `data-issue-link=...` and `data-query-params=...`, and the Vue can get it by `getAttribute`

wxiaoguang commented on templates/repo/diff/box.tmpl:
```diff
@@ -31,12 +31,38 @@
 	{{if not .DiffNotAvailable}}
+		{{if and .IsShowingOnlySingleCommit .PageIsPullFiles}}
+			<div class="ui info message">
+				<div class="text">{{.locale.Tr "repo.pulls.showing_only_single_commit" (ShortSha .BeforeCommitID)}}</div>
```
What's the purpose of `<div class="text">`?

sebastian-sauer replied:
removed - thanks

wxiaoguang commented on web_src/css/repo.css:
```diff
@@ -3350,3 +3350,24 @@ tbody.commit-list {
+  line-height: 1.4;
+}
+
+.ui.dropdown .menu.commit-selector-menu > .divider {
+  margin: 0;
+}
```
If we do not need the "margin" for a divider, then maybe there is no need to use "divider", add a "gt-border-top" to the items?

sebastian-sauer replied:
yes - removed it and added the border-top

wxiaoguang commented on web_src/js/components/DiffCommitSelector.vue:
```diff
@@ -0,0 +1,121 @@
+            <div class="gt-ellipsis text light-2">
+              {{ commit.committerOrAuthorName }}
+              <span class="text right">
+                <relative-time class="time-since" prefix="" :datetime="commit.time" data-tooltip-content data-tooltip-interactive="true">{{ commit.time }}</relative-time>
```
empty `data-tooltip-content`?

wxiaoguang commented on web_src/js/components/DiffCommitSelector.vue:
```diff
@@ -0,0 +1,121 @@
+  },
+  mounted() {
+    // fetch commit info
+    this.fetchCommits();
```
Only do "fetch" when the user wants to see the list (click the button)?

sebastian-sauer commented:
> > regarding the 500 on mysql:
> > Looks like there is a bug in FindLatestReviews (already existing function - introduced in #24796)
> > the max ( ) should be max() for mysql (without the space)
> 
> would it be worth to backport a fix for this only to 1.20?

I think this was introduced only on main - at least i couldn't find the method in 1.20 (https://github.com/go-gitea/gitea/blob/release/v1.20/models/issues/review.go)

sebastian-sauer commented:
all your review comments are now adressed (hopefully ;) ) and the hint for commit range selection via shift has been added. 

Are there any more comments / feedback? Thanks

puni9869 commented:
> Sorry for asking again - is there a chance to get this reviewed/merged or is there something else I need to add/modify?

Please wait, until re-reviewed. Lot of files has been changed. This may take lil more time to review. No need to sorry, you are fine. 😄

KN4CK3R commented on routers/web/repo/pull.go:
```diff
@@ -691,6 +691,103 @@ func PrepareViewPullInfo(ctx *context.Context, issue *issues_model.Issue) *git.C
+		ctx.JSON(http.StatusInternalServerError, err)
+		return
+	}
+	var prInfo *git.CompareInfo
+	var err error
+	if pull.HasMerged {
+		prInfo, err = baseGitRepo.GetCompareInfo(pull.BaseRepo.RepoPath(), pull.MergeBase, pull.GetGitRefName(), true, false)
+	} else {
+		prInfo, err = baseGitRepo.GetCompareInfo(pull.BaseRepo.RepoPath(), pull.BaseBranch, pull.GetGitRefName(), true, false)
+	}
+
```
```suggestion
	baseBranch := pull.BaseBranch
	if pull.HasMerged {
		baseBranch = pull.MergeBase
	}
	prInfo, err := baseGitRepo.GetCompareInfo(pull.BaseRepo.RepoPath(), baseBranch, pull.GetGitRefName(), true, false)
```

KN4CK3R commented on routers/web/repo/pull.go:
```diff
@@ -691,6 +691,103 @@ func PrepareViewPullInfo(ctx *context.Context, issue *issues_model.Issue) *git.C
+	}
+
+	resp := &pullCommitList{}
+	commits := make([]PullCommitInfo, len(prInfo.Commits))
+
+	for i := range prInfo.Commits {
+		commit := prInfo.Commits[i]
+
+		var committerOrAuthorName string
+		var time string
+		if commit.Committer != nil {
+			committerOrAuthorName = commit.Committer.Name
+			time = commit.Committer.When.String()
+		} else {
+			committerOrAuthorName = commit.Author.Name
+			time = commit.Author.When.String()
+		}
+
+		commits[i] = PullCommitInfo{
+			Summary:               commit.Summary(),
+			CommitterOrAuthorName: committerOrAuthorName,
+			ID:                    commit.ID.String(),
+			Time:                  time,
+		}
+	}
```
```suggestion
	commits := make([]PullCommitInfo, 0, len(prInfo.Commits))

	for _, commit := range prInfo.Commits {
		var committerOrAuthorName string
		var time string
		if commit.Committer != nil {
			committerOrAuthorName = commit.Committer.Name
			time = commit.Committer.When.String()
		} else {
			committerOrAuthorName = commit.Author.Name
			time = commit.Author.When.String()
		}

		commits = append(commits, PullCommitInfo{
			Summary:               commit.Summary(),
			CommitterOrAuthorName: committerOrAuthorName,
			ID:                    commit.ID.String(),
			Time:                  time,
		})
	}
```

KN4CK3R commented on routers/web/repo/pull.go:
```diff
@@ -754,6 +851,33 @@ func ViewPullFiles(ctx *context.Context) {
+		foundEndCommit := len(specifiedEndCommit) == 0
+
+		if !(foundStartCommit && foundEndCommit) {
+			for i := range prInfo.Commits {
+				if prInfo.Commits[i].ID.String() == specifiedStartCommit {
+					foundStartCommit = true
+				}
+				if prInfo.Commits[i].ID.String() == specifiedEndCommit {
+					foundEndCommit = true
+				}
+
+				if foundStartCommit && foundEndCommit {
+					break
+				}
+			}
```
```suggestion
			for _, commit := range prInfo.Commits {
				if commit.ID.String() == specifiedStartCommit {
					foundStartCommit = true
				}
				if commit.ID.String() == specifiedEndCommit {
					foundEndCommit = true
				}

				if foundStartCommit && foundEndCommit {
					break
				}
			}
```

wxiaoguang commented on web_src/js/components/DiffCommitSelector.vue:
```diff
@@ -0,0 +1,149 @@
+  methods: {
+    /** Opens our menu, loads commits before opening */
+    async toggleMenu() {
+      // load our commits when the menu is not yet visible (it'll be toggled after loading)
+      // and we got no commits
+      if (this.commits.length === 0 && this.menuVisible === false) {
+        await this.fetchCommits();
+      }
+      this.menuVisible = !this.menuVisible;
```
If the network/response is not fast (eg: 2second), then when a user clicks the dropdown, they sees nothing in 2 seconds.

Ideally, when the user clicks the dropdown, the menu should be shown immediately.

sebastian-sauer replied:
i'll add a loading animation inside the dropdown.

silverwind commented:
@sebastian-sauer did you remove the max width I set earlier? This is definietly too wide at over 800px. GitHub sets these dropdowns to static width of 380px.

<img width="1263" alt="image" src="https://github.com/go-gitea/gitea/assets/115237/40462dde-a8ab-46b0-9751-818a7aa5e3ea">

sebastian-sauer commented:
Sorry @silverwind - if i did so, that was not intended. I'll check and readd the max-width

silverwind commented:
One more thing: There is a double `border-top` issue coming from the `.menu > .menu` nesting, you should remove one of these `.menu`.

<img width="921" alt="image" src="https://github.com/go-gitea/gitea/assets/115237/6e9d5122-e055-4486-863d-beaaebd7bb04">

silverwind commented:
> Sorry @silverwind - if i did so, that was not intended. I'll check and readd the max-width

Thanks, I recommend something like `width: min(380px, 96vw)`.

sebastian-sauer commented:
> > Sorry @silverwind - if i did so, that was not intended. I'll check and readd the max-width
> 
> Thanks, I recommend something like `width: min(380px, 96vw)`.

I've added the max-width and reused shortsha and removed the double border-top

silverwind commented:
Hmm, it does give a 500 now for the SHA thingy:

<img width="1258" alt="image" src="https://github.com/go-gitea/gitea/assets/115237/8301d126-32c5-4790-9ff9-bad747dcda4a">

sebastian-sauer commented:
> Hmm, it does give a 500 now for the SHA thingy:
> 
> <img alt="image" width="1258" src="https://user-images.githubusercontent.com/115237/253063104-8301d126-32c5-4790-9ff9-bad747dcda4a.png">

Strange - just clicked through all urls and still working in my environment.
Which url did you use to get this error?

silverwind commented:
Trying again today, it seems to work. Double border is still an issue, checking...

silverwind commented:
Pushed some CSS tweaks:

<img width="554" alt="image" src="https://github.com/go-gitea/gitea/assets/115237/8c8d39e1-e006-400e-924d-6a0120b5dc5d">

What happened to the second line there? Previous screenshots showed `<user> 5 days` ago but now it only shows a ISO timstamp?

sebastian-sauer commented:
Sorry silverwind, I'll recheck this. Probably this broke during refactoring. Thanks!

sebastian-sauer commented:
Rechecked the dropdown and fixed the display of committer/author. This also fixed the relative-time component:

![image](https://github.com/go-gitea/gitea/assets/1135157/3ecb5ab2-2125-4a3c-b85d-6d911925b951)

delvh commented on tests/integration/pull_diff_test.go:
```diff
@@ -0,0 +1,58 @@
+	doTestPrDiff(t, "/user2/commitsonpr/pulls/1/commits/c5626fc9eff57eb1bb7b796b01d4d0f2f3f792a2", true, []string{"test3.txt"})
+}
+
+func TestPullDiff_CommitRangePRDiff(t *testing.T) {
+	doTestPrDiff(t, "/user2/commitsonpr/pulls/1/files/4ca8bcaf27e28504df7bf996819665986b01c847..23576dd018294e476c06e569b6b0f170d0558705", true, []string{"test2.txt", "test3.txt", "test4.txt"})
+}
+
+func TestPullDiff_StartingFromBaseToCommitPRDiff(t *testing.T) {
+	doTestPrDiff(t, "/user2/commitsonpr/pulls/1/files/c5626fc9eff57eb1bb7b796b01d4d0f2f3f792a2", true, []string{"test1.txt", "test2.txt", "test3.txt"})
+}
```
Should we perhaps move these routes to `{user}/{repo}/pulls/{id}/commits/{commit1}(..{commit2})`?
I don't have a strong preference towards either, but perhaps it makes it clearer for users that they are not reviewing the full PR at the moment, but only a snapshot

sebastian-sauer replied:
The URLs match the three urls GitHub uses for this feature - so I'd suggest we align our urls with GitHub. 
Is that ok?

delvh commented on tests/integration/pull_diff_test.go:
```diff
@@ -0,0 +1,58 @@
+	doTestPrDiff(t, "/user2/commitsonpr/pulls/1/files/c5626fc9eff57eb1bb7b796b01d4d0f2f3f792a2", true, []string{"test1.txt", "test2.txt", "test3.txt"})
+}
+
+func doTestPrDiff(t *testing.T, prDiffURL string, reviewBtnDisabled bool, expectedFilenames []string) {
```
There are differing opinions on this, but I think `doTestPRDiff` matches our convention the best.

sebastian-sauer replied:
changed the case of the R

delvh commented on tests/integration/pull_diff_test.go:
```diff
@@ -0,0 +1,58 @@
+	doc := NewHTMLParser(t, resp.Body)
+
+	// Assert all files are visible.
+	fileContents := doc.doc.Find("div.file-content")
```
```suggestion
	fileContents := doc.doc.Find(".file-content")
```
I don't like relying on specific elements, that breaks to quickly.

sebastian-sauer replied:
done.

delvh commented on web_src/js/components/DiffCommitSelector.vue:
```diff
@@ -0,0 +1,191 @@
+  mounted() {
+    // Add a click listener to close this menu on click outside of this element
+    //  when the dropdown is currently visible opened
+    document.body.addEventListener('click', (event) => {
+      if (this.$el.contains(event.target)) return;
```
Isn't that
```suggestion
    this.$el.addEventListener('click', (event) => {
```
?

sebastian-sauer replied:
No - this code checks for any click outside of the commitselector. Your suggestion would listen to all clicks on the element.
Please see the code comment above this line

(same logic is used in RepoBranchTagSelector.vue)

delvh replied:
Ah, I read that statement the wrong way around.
I read it as `only do something when the element was clicked` instead of `only do something when the element was`**not**` clicked`.

delvh commented on web_src/js/components/DiffCommitSelector.vue:
```diff
@@ -0,0 +1,191 @@
+      const results = await resp.json();
+      this.commits.push(...results.commits);
+      this.commits.reverse();
+      this.lastReviewCommitSha = results.last_review_commit_sha !== '' ? results.last_review_commit_sha : null;
```
```suggestion
      this.lastReviewCommitSha = results.last_review_commit_sha ? results.last_review_commit_sha : null;
```

sebastian-sauer replied:
hm - when changing the code i get an eslint warning to prefer logical operator over ternary. https://github.com/sindresorhus/eslint-plugin-unicorn/blob/v47.0.0/docs/rules/prefer-logical-operator-over-ternary.md

so i switched to a || instead.

delvh commented on web_src/js/components/DiffCommitSelector.vue:
```diff
@@ -0,0 +1,191 @@
+        // find all selected commits and generate a link
+        if (this.commits[0].selected) {
+          // first commit is selected - generate a short url with only target sha
+          const lastCommit = this.commits.findLast((x) => x.selected);
```
Should we perhaps add an extra check that redirects to `/files` when you select **all** commits?
While the current behavior is completely fine, I think it would be pretty user-friendly to redirect to the normal review page:
Evidently, the user wants to see every change and in other words should be able to approve it.

sebastian-sauer replied:
nice idea - i'll add this check in the component.

sebastian-sauer replied:
done.

delvh commented on web_src/js/components/DiffCommitSelector.vue:
```diff
@@ -0,0 +1,191 @@
+  <div class="ui dropdown custom">
+    <button
+      class="ui basic button"
+      @click.stop="toggleMenu()" @keyup.enter="toggleMenu()"
```
Not sure if `keyup.enter` suffices/ if we need it.
There should be a global JS selector that acts on `ui button` to redirect keypresses (both <kbd>Enter</kbd> and <kbd>Space</kbd>) to a button click.

sebastian-sauer replied:
Thanks - that was a bug. I removed the keyup.enter and now keyup opens the selector as expected.

delvh commented on routers/web/web.go:
```diff
@@ -1280,14 +1280,19 @@ func registerRoutes(m *web.Route) {
 			m.Group("/files", func() {
-				m.Get("", context.RepoRef(), repo.SetEditorconfigIfExists, repo.SetDiffViewStyle, repo.SetWhitespaceBehavior, repo.SetShowOutdatedComments, repo.ViewPullFiles)
+				m.Get("", context.RepoRef(), repo.SetEditorconfigIfExists, repo.SetDiffViewStyle, repo.SetWhitespaceBehavior, repo.SetShowOutdatedComments, repo.ViewPullFilesForAllCommitsOfPr)
+				m.Get("/{sha}", context.RepoRef(), repo.SetEditorconfigIfExists, repo.SetDiffViewStyle, repo.SetWhitespaceBehavior, repo.SetShowOutdatedComments, repo.ViewPullFilesForRangeOrStartingFromCommit)
```
Separate that into two routes and add restrictions on the accepted values (`{sha:[a-f0-9]{7,40}}`)

sebastian-sauer replied:
done.

silverwind commented on options/locale/locale_en-US.ini:
```diff
@@ -1659,6 +1659,13 @@ pulls.switch_comparison_type = Switch comparison type
 pulls.no_results = No results found.
+pulls.show_all_commits = Show all commits
+pulls.show_changes_since_your_last_review = Show changes since your last review
+pulls.showing_only_single_commit = Showing only changes of commit %[1]s
+pulls.showing_specified_commit_range = Showing only changes between %[1]s..%[2]s
```
GitHub just shows "Showing changes from n commit(s)", which I think it also sufficient. The user already had selected the commits by title, so a short message is sufficient to not distort the UI with long messages.

sebastian-sauer replied:
i personally like the commit sha display - but if you want this to be changed, i can change it.

Do you want me to change it to only show the number of commits?

delvh replied:
No, no, absolutely fine as it is.

delvh replied:
but perhaps we should show how many commits this range consists of.

sebastian-sauer replied:
hm - not sure. this would extend the text and the provided information is (at least in my opinion) not that important.

sebastian-sauer commented on routers/web/repo/pull.go:
```diff
@@ -694,6 +694,100 @@ func PrepareViewPullInfo(ctx *context.Context, issue *issues_model.Issue) *git.C
 	return compareInfo
 }
 
+type PullCommitInfo struct {
+	Summary               string `json:"summary"`
+	CommitterOrAuthorName string `json:"committer_or_author_name"`
+	ID                    string `json:"id"`
+	ShortSha              string `json:"short_sha"`
+	Time                  string `json:"time"`
+}
+
+type pullCommitList struct {
+	Commits             []PullCommitInfo  `json:"commits"`
+	LastReviewCommitSha string            `json:"last_review_commit_sha"`
+	Locale              map[string]string `json:"locale"`
+}
+
+// GetPullCommits get all commits for given pull request
+func GetPullCommits(ctx *context.Context) {
```
Do you mean moving all the structs to services and most of the logic of GetPullCommits?
Or just the structs? Or should i only extract getting the pullCommitInfo calculation and struct?

delvh replied:
Basically whatever you can extract.

sebastian-sauer replied:
I've extracted most of the logic that is not directly web related into services. Hope this is what you wanted.

delvh reviewed (commented):
Okay, the code is fine now.
Now on to the usability test.
<details><summary>These are my preliminary findings:</summary>

1. We need to visually separate the `Select commit` text from everything else. At first glance, it looks like one of the other commits instead of a help text…
![grafik](https://github.com/go-gitea/gitea/assets/51889757/03c9dd8f-5f8d-40fb-8d08-bf5379b6010e)
There are multiple options to achieve that: Different font, bigger/smaller margin/padding, divider. I have no immediate preference what would be the best fix.
2. The timestamps are buggy, hovering over them doesn't show the tooltip (even though according to the code they should - if I had to guess it's because they are added dynamically and not on initial page load)
3. I'm not sure, but the timestamps have the most visual weight for small commit titles (see screenshot). We should fix that somehow, i.e. making the commit titles bigger or the timestamps smaller. We could omit for example the duplicate timezone (whyever that is present?) And why do they look differently compared to every other timestamp?
4. That dropdown is completely inaccessible at the moment for some reason?
I can open it, but then I can't interact with it via the keyboard. At least I can close it again by clicking on another focused element again.
5. The way we currently set the commit range to use is somewhat counter-intuitive: The way it is described, I would have assumed that pressing <kbd>Shift</kbd>, then clicking and dragging the mouse, and then letting go of <kbd>Shift</kbd> would select all commits in between. While I can also work with your approach, there is an indication missing when we enter the selection phase. Perhaps we can do something similar to GitHub in that once we have selected our base commit, all commits between our mouse cursor and that base commit are highlighted?
6. Speaking of which, GitHub cancels the range selection when you let go of <kbd>Shift</kbd>. With your approach, I have no way to cancel an accidentally started range-selection. We can think about whether we want something like that as well.
7. The message 
![grafik](https://github.com/go-gitea/gitea/assets/51889757/ec57a0a4-12fc-4359-a9fb-5e4adad6aae6)
is all nice and dandy, however, I'm somewhat missing a link such as `Show all changes instead` in this message.
8. I think the dropdown is a bit misplaced. I didn't find it at first **and I was looking for it**. We've conditioned our users to see this corner (except for the `review` button) as dead-weight for options you rarely need. How are they supposed to find this functionality if it is that hidden that even I could not find it without looking at your screenshot closely?
9. It is a bit counter-intuitive that the `show changes since your last review` button is completely hidden **when the review happened on the latest commit** which makes it seem like I haven't reviewed this PR at all yet. I think it would be a better idea to disable it if that's the case. Note: It may still stay hidden when there was no review yet.
10. I've managed to break your `show changes since last review` functionality: 
![grafik](https://github.com/go-gitea/gitea/assets/51889757/e94bac39-6683-4ea2-ac21-e1227ff5cf87)
To do that, simply post a review on a commit, and then force-push another commit replacing the commit that was reviewed or one of its predecessors before reloading the page and clicking on the link. I don't know either if/think that we can catch this case somehow. At least the menu produces somewhat understandable output in this case:
![grafik](https://github.com/go-gitea/gitea/assets/51889757/eda72f64-1131-471f-899f-c4cbcf8ab31e)

Not all these things must necessarily be done in this PR, but `1` and `4` are mandatory, and I recommend `5`-`7`. The rest can also be fixed later on.
</details>

silverwind commented:
For 1, add `text light-2` or even `text light-3` class to the help text.

sebastian-sauer commented:
Thanks for your feedback. I'll check and try to fix your suggestions asap.

Regarding the timestamps - at least in my browser they worked yesterday evening. (i did recheck them).

wxiaoguang commented on web_src/js/components/DiffCommitSelector.vue:
```diff
@@ -0,0 +1,196 @@
+    width: 100%;
+  }
+
+  .ui.dropdown .menu .menu.is-loading {
+    left: 0;
+    height: 200px;
+    width: 200px !important;
+  }
+
+  .ui.dropdown .menu .menu.is-loading::after {
+    display: block !important; /* to override fomantic rule .ui.dropdown .menu .menu:after {display: none} */
+  }
```
I think it could avoid conflicting with Fomantic dropdown styles, use a separate element for "is-loading".

ps: there are some other nits in my mind, if you don't mind, could I edit your PR directly?

https://github.com/go-gitea/gitea/blob/79995a8194235ee994a1ce3cbe5eaa11358db885/web_src/js/components/RepoBranchTagSelector.vue#L41

sebastian-sauer replied:
Please feel free to edit the pr directly!

wxiaoguang replied:
> maybe you got an idea / solution for the inaccessibiltity with a keyboard?

I have thought about it. IMO it needs a separate framework level support in Vue. At the moment, Gitea's frontend lacks an a11y friendly Vue UI framework.

sebastian-sauer commented:
Just had a quick look at 2: looks like the relatve time component doesn't work in firefox. In Chrome it works perfectly - in FF it's not working.

The format of the timestamp was not valid for Firefox. I changed it to a well defined format and now it's working in FF and chrome

silverwind commented:
Indeed, I recall Spidermonkey's date parser is stricter than Blink. I think in this case Blink is actually in error but they are probably too afraid to change it to be standards-conform.

sebastian-sauer commented:
@delvh isn't 1 & 3 now resolved as well? 
The click message is in a light colour and the timestamps don't take that much space as they are displayed as they should :)

silverwind commented:
Just for the record, this is the date issue. We should ensure to never communicate bare `time` strings to browsers in the future.

```
$ eshost -e "new Date('2009-11-10 23:00:00 +0000 UTC')"
#### JavaScriptCore
Invalid Date

#### spidermonkey
Invalid Date

#### v8
Wed Nov 11 2009 00:00:00 GMT+0100 (Central European Standard Time)
```

delvh commented:
True, didn't notice it, but you are absolutely correct.
I've edited the message now.

delvh commented:
> $ eshost -e "new Date('2009-11-10 23:00:00 +0000 UTC')"

How is that invalid? Because of the `+timezone timezone`?

silverwind commented:
Related: https://github.com/github/relative-time-element/issues/266

sebastian-sauer commented:
@delvh i just added a new commit that makes the dropdown usable with keyboard navigation. At least in Chrome and FF this looks good to me - but i'd really appreciate your feedback on this.

sebastian-sauer commented:
https://github.com/go-gitea/gitea/pull/25528/commits/172418185895393502d2b0802293b7e8d3a2d9e6 adds a "Show all commits" link to the showing only commits message:

![image](https://github.com/go-gitea/gitea/assets/1135157/f112aa29-cc93-4d9c-89de-9d7c8db8035f)

![image](https://github.com/go-gitea/gitea/assets/1135157/3c8b5399-385c-45ad-bbad-5b77d158e15e)

This fixes 7 from @delvh's findings

sebastian-sauer commented:
And commit https://github.com/go-gitea/gitea/pull/25528/commits/f7acf4a986eedbadf7aa17680614261275c49ad7 adds an improved commit range selection (hopefully very close to the github one ;) )
This fixes point 5 + 6 from the list

![range-selection-improved](https://github.com/go-gitea/gitea/assets/1135157/ceee6aa3-087f-4c0b-9a7d-71251ffd5d05)

delvh commented on web_src/js/components/DiffCommitSelector.vue:
```diff
@@ -2,14 +2,20 @@
       @click.stop="toggleMenu()"
       :data-tooltip-content="locale.filter_changes_by_commit"
+      aria-haspopup="true"
+      tabindex="0"
```
I don't think that's necessary, and should be set automatically?

delvh commented on web_src/js/components/DiffCommitSelector.vue:
```diff
@@ -18,17 +24,17 @@
         </div>
         <!-- only show the show changes since last review if there is a review AND we are commits ahead of the last review -->
-        <div v-if="lastReviewCommitSha != null && commitsSinceLastReview > 0" class="vertical item gt-df gt-fc gt-gap-2 gt-border-secondary-top" @click="changesSinceLastReviewClick()">
+        <div v-if="lastReviewCommitSha != null" role="menuitem" tabindex="-1" class="vertical item gt-df gt-fc gt-gap-2 gt-border-secondary-top" :class="{disabled: commitsSinceLastReview === 0}" @keydown.enter="changesSinceLastReviewClick()" @click="changesSinceLastReviewClick()">
```
Instead of simulating buttons, we might as well use `<button>`s directly (if we replace the inner `<div>`s with `<span>`s).

delvh commented on web_src/js/components/DiffCommitSelector.vue:
```diff
@@ -85,11 +91,67 @@ export default {
+      const indexSelected = this.commits.findIndex((x) => x.selected);
+      const indexCurrentElem = this.commits.findIndex((x) => x.id === commit.id);
+      for (const [idx, commit] of this.commits.entries()) {
+        if (idx >= Math.min(indexSelected, indexCurrentElem) && idx <= Math.max(indexSelected, indexCurrentElem)) {
+          commit.hovered = true;
+        } else {
+          commit.hovered = false;
+        }
```
```suggestion
        commit.hovered = idx >= Math.min(indexSelected, indexCurrentElem) && idx <= Math.max(indexSelected, indexCurrentElem);
```

delvh reviewed (approved):
Only some minor code comments left.
Other than that LGTM.

wxiaoguang commented:
Some changes:

1. Bypass the XORM bug for `max ( id )`
2. There were class "commit-selector-menu" and id "commit-selector-menu", it seems that only ID is enough, now only use `id="diff-commit-selector-menu"` (the prefix is used to prevent from conflicting from other IDs)
3. Use `loading-indicator` instead of battling with "menu::after"
4. Wrap some pretty long lines to multiple lines
5. Add `@click.ctrl.exact` and `@click.meta.exact`, it brings the consistent user experience if the user would like to open the link in a new window.

wxiaoguang commented on web_src/js/components/DiffCommitSelector.vue:
```diff
@@ -95,6 +95,20 @@ export default {
       }
     });
 
+    // Add an eventlistener for shift keyup
+    // only is active if hoverActivated is true.
+    document.addEventListener('keyup', (evt) => {
```
Why not reuse `this.$el.addEventListener('keyup', ...` ?

sebastian-sauer replied:
If someone does keyup outside of the element i wanted to catch this, too.

wxiaoguang replied:
Hmm ... still don't get the point how could it happen.

The focus should always be in the "$el", if the focus is not in the "$el", I think the "$el" should be hidden, then no need to handle the keyup?

sebastian-sauer replied:
Sounds reasonable - do you want to change it or should I do this (today evening German time)?

wxiaoguang replied:
Either is fine (I am going out soon, so if I have time half-day later I could also do the change)

sebastian-sauer commented:
Thanks @wxiaoguang for your changes - these look good to me!

delvh commented:
@sebastian-sauer I've updated your screenshots now with the current state.

delvh commented on templates/repo/diff/box.tmpl:
```diff
@@ -101,8 +101,8 @@
 					{{$isExpandable := or (gt $file.Addition 0) (gt $file.Deletion 0) $file.IsBin}}
-					{{$isReviewFile := and $.IsSigned $.PageIsPullFiles (not $.IsArchived) $.IsShowingAllCommits}}
-					<div class="diff-file-box diff-box file-content {{TabSizeClass $.Editorconfig $file.Name}} gt-mt-3" id="diff-{{$file.NameHash}}" data-old-filename="{{$file.OldName}}" data-new-filename="{{$file.Name}}" {{if or ($file.ShouldBeHidden) (not $isExpandable)}}data-folded="true"{{end}}>
+          {{$isReviewFile := and $.IsSigned $.PageIsPullFiles (not $.IsArchived) $.IsShowingAllCommits}}
```
Hmm… Somehow this line lost half of its indentation in the merge.

sebastian-sauer commented:
Thank you @delvh for resolving the merge conflict. And thanks to all of you for all your support! It's always fun contributing to gitea!

sebastian-sauer commented:
I'll check this and push a fix asap.

sebastian-sauer commented:
Fixed.

delvh commented:
I've just noticed something strange, which is present both in my new screenshots and in the original ones:
The count of added and deleted lines doesn't add up with the reality, especially when only reviewing a single commit of a PR:
My third screenshot for example shows `2 additions, 1 deletion`. At the same only `1 file changed`, and this file has `1 addition, 0 deletions`. How can that be?

sebastian-sauer commented:
hm - is there anything special with your commit? in my local environment it shows the exact number of lines changed.

![image](https://github.com/go-gitea/gitea/assets/1135157/57cdb206-6aaf-4ec9-b033-61bf25699cfc)

sebastian-sauer commented:
@delvh do you got Ignore Whitespace active? This could be the reason

delvh commented:
I mean, it doesn't appear to occur on every commit.
Let's just book it under `Cannot reproduce` and be happy.
We'll see it again when a bug report about is opened.

sebastian-sauer commented:
> I mean, it doesn't appear to occur on every commit. Let's just book it under `Cannot reproduce` and be happy. We'll see it again when a bug report about is opened.

Please see my last comment - maybe it's caused by ignore whitespace option?
